harfbuzz: cross compilation

Mindavi opened the pull request:
###### Motivation for this change

Support cross-compilation by disabling introspection. Tested by building for aarch64-multiplatform.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented on pkgs/development/libraries/harfbuzz/default.nix:
```diff
@@ -43,24 +57,28 @@ stdenv.mkDerivation {
     ++ lib.optionals withCoreText [ ApplicationServices CoreText ];
 
-  propagatedBuildInputs = []
+  propagatedBuildInputs = [ ]
     ++ optional withGraphite2 graphite2
```
```suggestion
  propagatedBuildInputs = optional withGraphite2 graphite2
```

Ericson2314 replied:
I actually kinda like this style, easier to rearrange lines later.

Mindavi replied:
Done.

sternenseemann commented:
Is it impossible to cross-compile `gobject-introspection` or with `gobject-introspection`-support? Or is it just not possible *in nixpkgs*. A comment explaining the situation would be helpful.

Mindavi commented:
In the future it might become possible due to the work of @Ericson2314 here: https://github.com/NixOS/nixpkgs/pull/88222. See also the related issue: https://github.com/NixOS/nixpkgs/issues/72868.

However, in the present day it's not possible and I think John Ericson is working hard to get the support upstream to prevent having to maintain a fork. It seems like upstream is not very willing to accept the patches though, so the process seems to be slow. As you can see in the PRs, the support is also not complete, so there's some work to do on that.

Any comments on this @Ericson2314 ? This is just my own interpretation of what's happening.

So in short: it's not possible right now, but might be in the future.

sternenseemann commented:
Ah, thanks for the explanation! Could you link the related issue or PR with a short note in a comment? I think this makes it clearer to the reader what is going on.

Mindavi commented:
I'd rather put it in the commit message to prevent adding that comment on all places that disable introspection (there are at least ~15 or so). But either way is fine by me :).

sternenseemann commented:
Sure!

sternenseemann commented:
@ofborg build pkgsCross.aarch64-multiplatform.harfbuzz

Ericson2314 commented on pkgs/development/libraries/harfbuzz/default.nix:
```diff
@@ -1,13 +1,26 @@
+, CoreText
 , withCoreText ? false
 , withIcu ? false # recommended by upstream as default, but most don't needed and it's big
 , withGraphite2 ? true # it is small and major distros do include it
```
```suggestion
, withGraphite2 ? true # it is small and major distros do include it
, withIntrospection ? stdenv.buildPlatform == stdenv.hostPlatform
```

jtojnar replied:
Alternately,

you can define `withIntrospection` in a `let` block as not to clutter the inputs with stuff noone will override anyway.

Ericson2314 commented on pkgs/development/libraries/harfbuzz/default.nix:
```diff
@@ -43,8 +56,11 @@ stdenv.mkDerivation {
     (mesonFeatureFlag "graphite" withGraphite2)
     (mesonFeatureFlag "icu" withIcu)
     (mesonFeatureFlag "coretext" withCoreText)
+    (mesonFeatureFlag "introspection" (stdenv.buildPlatform == stdenv.hostPlatform))
```
```suggestion
    (mesonFeatureFlag "introspection" withIntrospection)
```

Ericson2314 commented on pkgs/development/libraries/harfbuzz/default.nix:
```diff
@@ -58,9 +74,10 @@ stdenv.mkDerivation {
   buildInputs = [ glib freetype cairo ] # recommended by upstream
-    ++ lib.optionals withCoreText [ ApplicationServices CoreText ];
+    ++ lib.optionals withCoreText [ ApplicationServices CoreText ]
+    ++ lib.optionals (stdenv.buildPlatform == stdenv.hostPlatform) [ gobject-introspection ];
```
```suggestion
    ++ lib.optionals withIntrospection [ gobject-introspection ];
```

Mindavi replied:
It seems that you and @jtojnar disagree on the style here. What shall I do here? I personally find the `withIntrospection` nice since it's also easy to search for if the introspection stuff gets updated to support cross.

But I don't have a strong opinion on what's best.

jtojnar replied:
Yeah, I am trying to reduce the noise in the expressions I maintain – the chance that someone will want to disable introspection aside from cross-compiling is pretty small so the additional clutter is not worth it IMO. But maintainers of each package will have to choose.

Ericson2314 replied:
@jtojnar says let bind it rather than make it a parameter. I don't care strongly about that; I just want a variable to be somehow bound (let or parameter) with a descriptive name, rather than repeating the cross check which isn't inherently related to introspection.

Mindavi replied:
Thanks for the tip! Updated with a let-expression.

Ericson2314 commented:
Yes, I am blocked for gobject-introspection on upstream merging PRs for like a year now: https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/227 and https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/224. They even approved the first one. No idea what the deal is; it's been frustrating.

Mindavi committed:
harfbuzz: nixpkgs-fmt

Mindavi committed:
harfbuzz: support cross-compilation by disabling introspection

See also discussion about this in the WIP PR from @Ericson2314:
- https://github.com/NixOS/nixpkgs/pull/88222

Related issue:
- https://github.com/NixOS/nixpkgs/issues/72868

Mindavi committed:
harfbuzz: style cleanup

Mindavi commented:
Updated to address all comments. Glad we could find a nice solution. Thanks for reviewing! (and great to see that some stuff from Ericson2314 has finally been merged upstream on gobject-introspection, great step in the right direction :smile:).

Mindavi commented on pkgs/development/libraries/harfbuzz/default.nix:
```diff
@@ -1,29 +1,43 @@
   inherit (lib) optional optionals optionalString;
   mesonFeatureFlag = opt: b:
     "-D${opt}=${if b then "enabled" else "disabled"}";
+  withIntrospection = stdenv.buildPlatform == stdenv.hostPlatform;
```
Fine with me.
